### PR TITLE
BUG: sparse: undesired behaviour of 1D/2D matmul

### PR DESCRIPTION
**Pull request by anushkasuyal**
A per the [array API standards for matmul](https://data-apis.org/array-api/2023.12/API_specification/generated/array_api.matmul.html), a 1-D array with shape `(N,)` multiplied by a 2-D array with shape `(N,K)` would give a 1-D array with shape `(K,)`. For K=1, we get as output a 1-D array of one element. The current definition of the method `_matmul_dispatch` gives a 0-D output for `matmul` of a sparse `(N,)` array with a dense `(N,1)` array, and it gives a `(1,1)` shape output for `matmul` of a sparse `(N,)` array with a dense `(N,1)` list.

This PR rectifies this undesired behaviour to ensure the array API standards are followed.

### CODE REVIEW

File: scipy/sparse/tests/test_common1d.py [hunk 1]
```diff
@@ -352,10 +352,8 @@ def test_matvec(self, spcreator):
         assert (A @ np.array([1, 2, 3])).shape == ()
         assert Asp @ np.array([1, 2, 3]) == 11
         assert (Asp @ np.array([1, 2, 3])).shape == ()
-        assert (Asp @ np.array([[1], [2], [3]])).shape == ()
```
Comment (dschult) [inline scipy/sparse/tests/test_common1d.py:355]: Can we leave these two tests in (but change the resulting shape to `(1,)`?

Actually, they look like exactly the same test. So you can delete the last one, but can you keep the first one with a corrected result?  I tried it locally and it worked.

**Review by dschult — APPROVED** [review]
This fixes a discovered bug in the 1D matmul machinery where the return value did not have the right shape. (1D sparse @ 2D dense column should be 1D, not 0D).

It fixes a test in `test_base.py`, removes a duplicated version of the same test, and adds tests in `test_coo.py`

I approve this PR.  It is ready to merge

**Comment by dschult** [comment]
@perimosocordiae if you get a chance, this is a good PR to review. I made a mistake in reshaping one of the cases for 1D matmul (1D sparse @ 2D dense). @anushkasuyal found the bug and fixed it. :)

**Review by perimosocordiae — APPROVED** [review]

**Comment by perimosocordiae** [comment]
Thanks @anushkasuyal for the fix!

**Pull request merged**